### ISSUE TITLE
Solve use_dll-incompatibility problem on Windows in master

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#1227]
After merging the `temporary_terms`-branch (https://github.com/DynareTeam/dynare/commit/127637ffd672fdd8143ee13f2a38234e597d94d3) the `c`-files created are not `C89`-compatible, but require `C99`. This is a problem for Windows machines as Visual Studio before version 2015 only supported `C89`, resulting in errors of the form
`error C2143: syntax error : missing ';' before 'type'`
Before Matlab version `R2014b` users could use the `cygwin`-option by using the provided `mexopts-win64.bat` from http://www.dynare.org/DynareWiki/ConfigureMatlabWindowsForMexCompilation, which should not have this restriction. Since `R2015b`, `mingw` is available, which also supports `C99` (#1226). Moreover, Visual Studio 2015 finally supports `C99` and has a free available version that is supported since `R2015b`. That leaves 64bit Matlab-versions between `R2014a` and  `R2015a` unable to deal with the problem unless someone figures out how to solve #641 for `cygwin` (potentially following the advice on how to set up the `xml`-file at http://stackoverflow.com/questions/8552580/using-gcc-mingw-as-matlabs-mex-compiler). Given that there seems to be not that much demand for `use_dll` on Windows, I would propose for 4.5 to focus on the `mingw`-support and state the above restrictions clearly in the release notes and the manual

**Comment by MichelJuillard** [comment]
If this occurred after merging temporary_terms branch, this is a bug. There is no reason for the C files not to be C89 compatible

**Comment by JohannesPfeifer** [comment]
The problem is that after the change we have code like
```
  /* Residual equations */
double T14 = pow(y[7],1+params[5]);
lhs =y[4]*params[4]*T14;
rhs =(1-params[2])*y[3];
residual[0]= lhs-rhs;

  /* Jacobian  */
  if (g1 == NULL)
    return;

double T106 = T45*getPowerDeriv(y[0],params[2],1);
}
``` 

This is not valid `C89` because the temporary term `T106` is only created when the Jacobian is requested, but in `C89` all variables must be defined before other code.

**Comment by MichelJuillard** [comment]
So we need to break it in parts:

a residual function 
a Jacobian function
a Hessian function
a 3rd order derivative function
and a main function calling all four other functions

**Comment by houtanb** [comment]
@MichelJuillard the problem with multiple functions is that we would need to return arrays of temporary variables that have previously been calculated. So, the residual function would need to return the temporary variables it calculated so that the Jacobian could use these, and so on. 

Another option is to move our mex files to C++.

@JohannesPfeifer can you try passing the `/TP` flag to the MSVC compiler? This will require other changes to our code, but I'm just looking to see if you'll still get these declaration errors or not. See https://msdn.microsoft.com/en-us/library/032xwy55.aspx

**Comment by JohannesPfeifer** [comment]
Changing Matlab's `xml`-file's `COMPFLAGS=` to include `/TP` solves the issue for msvc

**Comment by houtanb** [comment]
@JohannesPfeifer good. @MichelJuillard are you ok with this fix?

**Comment by MichelJuillard** [comment]
@houtanb OK for me

**Comment by houtanb** [comment]
#1228 fixes a bug in 83eeafa
